lib/pbench/server/api/resources/api_key.py

```diff
@@ -0,0 +1,104 @@
+from http import HTTPStatus
+
+from flask import jsonify, make_response
+from flask.wrappers import Request, Response
+from sqlalchemy.exc import IntegrityError, SQLAlchemyError
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    APIAbort,
+    ApiAuthorizationType,
+    ApiBase,
+    ApiContext,
+    APIInternalError,
+    ApiMethod,
+    ApiParams,
+    ApiSchema,
+)
+import pbench.server.auth.auth as Auth
+from pbench.server.database.models.api_key import APIKeyError, APIKeys
+from pbench.server.database.models.audit import (
+    Audit,
+    AuditStatus,
+    AuditType,
+    OperationCode,
+)
+
+
+class APIKey(ApiBase):
+    def __init__(self, config: PbenchServerConfig):
+        super().__init__(
+            config,
+            ApiSchema(
+                ApiMethod.POST,
+                OperationCode.CREATE,
+                audit_type=AuditType.API_KEY,
+                audit_name="apikey",
+                authorization=ApiAuthorizationType.NONE,
+            ),
+        )
+        self.server_config = config
+
+    def _post(
+        self, params: ApiParams, request: Request, context: ApiContext
+    ) -> Response:
+        """
+        Post request for generating a new persistent API key.
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+
+        Returns:
+            Success: 201 with api_key and user
+
+        Raises:
+            APIAbort, reporting "UNAUTHORIZED"
+            APIInternalError, reporting the failure message
+        """
+        user = Auth.token_auth.current_user()
+
+        if not user:
+            raise APIAbort(
+                HTTPStatus.UNAUTHORIZED,
+                "User provided access_token is invalid or expired token",
+            )
+        try:
+            new_key = APIKeys.generate_api_key(Auth, user=user)
+        except APIKeyError as e:
+            raise APIInternalError(e.message)
+        except Exception as e:
+            raise APIInternalError("Unexpected backend error") from e
+
+        try:
+            key = APIKeys(
+                api_key=new_key,
+                user=user,
+            )
+            key.add()
+        except IntegrityError:
```

Comment:
You actually shouldn't ever be able to get this, because `add` catches any `Exception` and raises `APIKeyError`. But as I said before, when SQLAlchemy raises an `IntegrityError` it could mean either that you've violated your primary key uniqueness constraint (duplicate) *or* that you've violated one of the column `nullable = False` constraints, and there's no way to tell without the `_decode` logic we've already established has been duplicated too many times.

Still, without that, we shouldn't be blindly claiming "duplicate api key"; and since you're doing this as an internal server error anyway just *say* there was a database error.

Note that a `raise [...] from` both here and on the `APIKeyError` in `add` would make sure that we have the whole stack to analyze when it happens.

Philosophically, though, a "duplicate" means the user tried to generate two API keys closely enough together that the timestamps don't differ to the resolution we're using, and calling that an "internal server error" seems really unfair. (Both to us and to the user.)